lib/events.js

```diff
@@ -75,15 +78,78 @@ EventEmitter.init = function() {
     }
   }
 
-  if (this._events === undefined ||
-      this._events === Object.getPrototypeOf(this)._events) {
-    this._events = Object.create(null);
-    this._eventsCount = 0;
-  }
+  if (this[kEvents] === Object.getPrototypeOf(this)[kEvents])
+    this[kEvents] = undefined;
 
   this._maxListeners = this._maxListeners || undefined;
 };
 
+// Events proxy for compatibility with older user-land code
+const proxyEventsHandler = {
+  deleteProperty({ emitter }, prop) {
+    const events = emitter[kEvents];
+    return events === undefined ? false : emitter[kEvents].delete(prop);
+  },
+  get({ emitter }, prop) {
+    let events = emitter[kEvents];
+    if (events === undefined)
+      events = emitter[kEvents] = new Map();
+    if (typeof prop !== 'string' && typeof prop !== 'symbol')
+      prop = String(prop);
+    return events.get(prop);
+  },
+  getOwnPropertyDescriptor({ emitter }, prop) {
+    const events = emitter[kEvents];
+    if (events === undefined)
+      return undefined;
+    const value = events.get(prop);
+    if (value === undefined)
+      return undefined;
+    return {
+      value,
+      writable: true,
+      configurable: true,
+      enumerable: true
+    };
+  },
+  getPrototypeOf() {
+    return null;
+  },
+  ownKeys({ emitter }) {
+    const events = emitter[kEvents];
+    if (events === undefined)
+      return [];
+    return [...events.keys()];
+  },
+  set({ emitter }, prop, value) {
+    let events = emitter[kEvents];
+    if (events === undefined)
+      events = emitter[kEvents] = new Map();
+    if (typeof prop !== 'string' && typeof prop !== 'symbol')
+      prop = String(prop);
+    events.set(prop, value);
+    return true;
+  }
+};
+
+Object.defineProperty(EventEmitter.prototype, '_events', {
+  configurable: true,
+  enumerable: true,
+  get() {
+    const proxy = this[kEventsProxy];
+    if (proxy === undefined) {
+      const scope = Object.create(null);
+      scope.emitter = this;
+      return this[kEventsProxy] = new Proxy(scope, proxyEventsHandler);
```

Comment:
Could you explain the purpose of `scope`? Maybe add a comment or so?

Reply:
Yeah, definitely. It basically has to do with https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy/handler/ownKeys

I don't know if it actually needs to be `Object.create(null)` though. I think it can just be a plain object.